File: Java/Stack/Implementaciones/Stack.java
package Stack.Implementaciones;

public class Stack<T> {
    private Node<T> top;

    public Stack() {
        this.top = null;
    }

    // Método para verificar si la pila está vacía
    public boolean isEmpty() {
        return top == null;
    }

    // Método para agregar un elemento a la pila
    public void push(T value) {
        Node<T> newNode = new Node<>(value);
        newNode.next = top;
        top = newNode;
    }

    // Metodo para mostrar la pila
    public static<T> void printList(Node<T> head) {
        Node<T> current = head;
        // recorrer hasta current sea null
        while (current != null) {
            System.out.print(current.data + " ");
            current = current.next;
        }
    }

    // Método para eliminar y retornar el elemento en la parte superior de la pila
    public T pop() {
        if (isEmpty()) {
            throw new IllegalStateException("Pila está vacía");
        }
        T value = top.data;
        top = top.next;
        return value;
    }
    
    // Método para obtener el valor del elemento en la parte superior de la pila
    public T peek() {
        if (isEmpty()) {
            throw new IllegalStateException("Pila está vacía");
        }
        return top.data;
    }

    // Método para obtener el tamaño de la pila
    public int size() {
        int count = 0;
        Node<T> current = top;
        while (current != null) {
            count++;
            current = current.next;
        }
        return count;
    }

    // Método para obtener la cabeza de la pila
    public Node<T> getTop() {
        return top;
    }

    public static void main(String[] args) {
        Stack<Integer> pila = new Stack<>();
        Boolean isEmpty = pila.isEmpty();
        System.out.print(isEmpty);
        System.out.println();
        pila.push(1);
        pila.push(2);
        pila.push(3);
        pila.push(4);
        pila.push(5);

        printList(pila.top);
        System.out.println();
        System.out.println("Elemento en la parte superior: " + pila.peek());
        System.out.println("Tamaño de la pila: " + pila.size());
        pila.pop();
        System.out.println("Después de hacer pop:");
        printList(pila.top);

    }

}

File: C++/EstructurasDeDatos/Stack/Implementaciones/SimpleStack/Stack.cpp
#include "Stack.h"
#define endl "\n"


Node::Node(int val){
    data = val;
    next = nullptr;
}

Stack::Stack(){
    top = nullptr;
    size = 0;
}

void Stack::push(int val){
    Node* newNode = new Node(val);
    if (top == nullptr){
        top = newNode;
        size++;
        return;
    } else {
        newNode -> next = top;
        top = newNode;
        size++;
        return;
    }
}

Node* Stack::pop(){
    if (top == nullptr){
        cout << "Nada para eliminar" << endl;
        return nullptr;
    }   
    Node* temp = top;
    top = top -> next;
    size--;
    return temp;
}

Node* Stack::peek(){
    if (top == nullptr){
        cout << "Empty" << endl;
        return nullptr;
    }
    return top;
}

bool Stack::is_empty(){
    if (top == nullptr)
        return true;
    return false;
}

int Stack::get_size(){
    return size;
}

void Stack::display(){
    Node* iter = top;
    while (iter != nullptr){
        cout << iter -> data << " ";
        iter = iter -> next;
    }
    return;
}


int main(){
    cout << boolalpha;
    Stack pila;
    cout << "Size inicial de la pila" << endl;
    cout << pila.get_size() << endl;
    cout << "Insertar elementos 1, 2 y 3" << endl;
    pila.push(1);
    pila.push(2);
    pila.push(3);
    cout << "Imprimir pila" << endl;
    pila.display();
    cout << endl;
    cout << "Eliminar el elemento top";
    cout << pila.pop() -> data << endl;
    cout << "Imprimir pila" << endl;
    pila.display();
    cout << endl;
    cout << "Elemento al tope de la pila" << endl;
    cout << pila.peek() -> data << endl;
    cout << "Mirar si está vacía la pila" << endl;
    cout << pila.is_empty() << endl;
    cout << "Tamaño de la pila:" << endl;
    cout << pila.get_size() << endl;
}
